### PR TITLE
GH-90699: Remove remaining `_Py_IDENTIFIER`  stdlib usage

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-90699

Automerge-Triggered-By: GH:ericsnowcurrently

### CODE REVIEW

File: Include/internal/pycore_global_strings.h [hunk 1]
```diff
@@ -21,7 +21,7 @@ extern "C" {
 #define STRUCT_FOR_STR(NAME, LITERAL) \
     STRUCT_FOR_ASCII_STR(LITERAL) _ ## NAME;
 #define STRUCT_FOR_ID(NAME) \
-    STRUCT_FOR_ASCII_STR(#NAME) _ ## NAME;
+    STRUCT_FOR_ASCII_STR(#NAME) __id__ ## NAME;
```
Comment (ericsnowcurrently) [inline Include/internal/pycore_global_strings.h]: Why is this necessary?

File: Modules/_testcapimodule.c [hunk 2]
```diff
@@ -3369,15 +3367,18 @@ slot_tp_del(PyObject *self)
     /* Save the current exception, if any. */
     PyErr_Fetch(&error_type, &error_value, &error_traceback);
 
+    PyObject *tp_del = PyUnicode_InternFromString("__tp_del__");
+    if (tp_del == NULL) {
+        return;
+    }
```
Comment (ericsnowcurrently) [inline Modules/_testcapimodule.c]: The error should be cleared and the current exception restored, right?

```suggestion
    if (tp_del == NULL) {
        PyErr_WriteUnraisable(NULL);
        PyErr_Restore(error_type, error_value, error_traceback);
        return;
    }
```

File: Modules/_datetimemodule.c [hunk 3]
```diff
@@ -5747,7 +5738,14 @@ datetime_repr(PyDateTime_DateTime *self)
 static PyObject *
 datetime_str(PyDateTime_DateTime *self)
 {
-    return _PyObject_CallMethodId((PyObject *)self, &PyId_isoformat, "s", " ");
+    PyObject *space = PyUnicode_FromString(" ");
```
Comment (ericsnowcurrently) [inline Modules/_datetimemodule.c:5741]: I'm surprised we don't have a `_Py_DECLARE_STR(space, " ")` somewhere (not that we need to). 😄

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]

File: Modules/_testcapimodule.c (hunk 2 above)
Comment (kumaraditya303) [reply Modules/_testcapimodule.c]: Right, I forgot push the fix for this yesterday, fixed now.

File: Include/internal/pycore_global_strings.h (hunk 1 above)
Comment (kumaraditya303) [reply Include/internal/pycore_global_strings.h]: This is necessary to avoid conflicts with Windows header globals and field names. See https://github.com/kumaraditya303/cpython/actions/runs/3387388754/jobs/5628092850 as example conflicting `_tzname`  with https://learn.microsoft.com/en-us/cpp/c-runtime-library/daylight-dstbias-timezone-and-tzname?view=msvc-170. I have clarified the comment w.r.t globals.

Comment (ericsnowcurrently) [reply Include/internal/pycore_global_strings.h]: Ah, thanks for clarifying.  From what I can tell, though, the problem is with macros rather than globals.

It looks like MSVC [*used to* have a global variable named `_tzname`](https://learn.microsoft.com/en-us/cpp/c-runtime-library/daylight-dstbias-timezone-and-tzname).  They later added a function called `_get_tzname()` (as well as a similar `__tzname()`) to do the same thing, deprecated `_tzname`, and then added something like `#define _tzname __tzname()`.  Thus, when we have `_tzname` as a field name, it gets expanded to `__tzname()` and we end up with "error C2032: '__tzname': function cannot be member of struct".

So it might be simpler to change the one outlier, using something like `_Py_DECLARE_STR(id_tzname, ")` and `_Py_STR(id_tzname)` (and a comment about why), instead of sticking with `_Py_ID()` and adding that `__id__` prefix.

---

Then again, the longer prefix does sort of future-proof other potential macro issues.  What you've done is probably the better way forward, so let's stick with that.  I recommend three changes, though:

* [ ] make the prefix even less likely to collide with macros in the future: `_py_`, `_py_id_`, or even `_py_str_` (my preference would be `_py_`)
* [ ] likewise adjust the prefix for `_Py_STR()`: `_py_`, or `_py_str_`, just to play it safe and stay consistent with `_Py_ID()`
* [ ] update the comment to `// to help avoid collisions with keywords, macros, etc.

FWIW, I think it would be best to have the same prefix for `_Py_global_strings.literals` and `_Py_global_strings.identifiers`, which would eliminate `_py_id_` as an option.  `_py_str_` is a bit redundant (and/or ambiguous) so I'd go with `_py_`.

Comment (kumaraditya303) [reply Include/internal/pycore_global_strings.h]: Okay let's go with `_py_` our tradition with underscore py :)

Comment (kumaraditya303) [reply Include/internal/pycore_global_strings.h]: Done in https://github.com/python/cpython/pull/99067/commits/729784a694fef3624709f3b87afc2b531123509b

Comment (kumaraditya303) [reply Include/internal/pycore_global_strings.h]: Thanks for the explanation.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

(Thanks for the changes.)

File: Modules/ossaudiodev.c [hunk 4]
```diff
@@ -542,9 +542,7 @@ oss_self(PyObject *self, PyObject *unused)
 static PyObject *
 oss_exit(PyObject *self, PyObject *unused)
 {
-    _Py_IDENTIFIER(close);
```
Comment (ericsnowcurrently) [inline Modules/ossaudiodev.c:545]: We also need to remove `#define NEEDS_PY_IDENTIFIER`, right?

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
One small thing and then LGTM.

Also, were you planning on removing `_Py_IDENTIFIER()` from Programs/_testembed.c?  (Presumably we'd drop `test_unicode_id_init()`.)  Then again, that should probably be part of the issue/PR that actually eliminates `_Py_IDENTIFIER()`.

File: Modules/ossaudiodev.c (hunk 4 above)
Comment (kumaraditya303) [reply Modules/ossaudiodev.c:545]: done

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
Refleak buildbots fail since this commit. Example:
https://buildbot.python.org/all/#/builders/217/builds/878

```
2 tests failed:
    test_finalization test_gc
```

Example:

```
$ make && ./python -m test -R 3:3 test_finalization -m test_legacy_self_cycle 
...
test_finalization leaked [1, 1, 1] references, sum=3
...
```

cc @pablogsal

**Comment by pablogsal** [comment]
Please take a look, otherwise we will need to revert if is not fixed in 24h per our buildbot policy.

**Comment by ericsnowcurrently** [comment]
If @kumaraditya303 doesn't get to it first, I'll tackle this first thing tomorrow morning (~15 hours from now).

**Comment by kumaraditya303** [comment]
https://github.com/python/cpython/pull/99236 fixes the refleak.

**Comment by vstinner** [comment]
The PR got merged, I confirm that it does fix the issue:

```
$ ./python -m test -R 3:3 test_gc test_finalization
(...)
Tests result: SUCCESS
```
